packages/webpack-cli/src/plugins/dotenv-webpack-plugin.ts

```diff
@@ -0,0 +1,200 @@
+import dotenv from "dotenv-defaults";
+import { Compiler, DefinePlugin } from "webpack";
+
+interface EnvVariables {
+  [key: string]: string;
+}
+
+interface Logger {
+  info(message: string): void;
+  warn(message: string): void;
+  error(message: string): void;
+}
+
+interface DotenvConfig {
+  paths?: string[];
+  prefixes?: string[];
+  systemvars?: boolean;
+  allowEmptyValues?: boolean;
+  expand?: boolean;
+  ignoreStub?: boolean;
+  defaults?: boolean | string;
+}
+
+type CodeValue = string;
+
+const interpolate = (env: string, vars: EnvVariables): string => {
+  const matches = env.match(/\$([a-zA-Z0-9_]+)|\${([a-zA-Z0-9_]+)}/g) || [];
+  matches.forEach((match) => {
+    env = env.replace(match, interpolate(vars[match.replace(/\$|{|}/g, "")] || "", vars));
+  });
+  return env;
+};
+
+const isMainThreadElectron = (target: string | undefined): boolean =>
+  !!target && target.startsWith("electron") && target.endsWith("main");
+
+export class Dotenv {
+  #logger!: Logger;
+  #config: DotenvConfig;
+  #inputFileSystem: any;
+  #cached: boolean;
+
+  constructor(config: DotenvConfig = {}) {
+    this.#config = {
+      paths: ["./.env"],
+      prefixes: ["process.env.", "import.meta.env."],
+      ...config,
+    };
+    this.#cached = false;
+  }
+
+  public apply(compiler: Compiler): void {
+    this.#inputFileSystem = compiler.inputFileSystem;
+    this.#logger = compiler.getInfrastructureLogger("dotenv-webpack-plugin");
+    compiler.hooks.thisCompilation.tap("dotenv-webpack-plugin", async (compilation) => {
+      const cache = compilation.getCache("dotenv-webpack-plugin-compiler");
+      const nodeEnv = process.env.NODE_ENV || compilation.options.mode;
+      if (this.#config.paths) {
+        this.#config.paths = [
+          ...this.#config.paths,
+          ...(nodeEnv === "test" ? [] : [".env.local"]),
+          `.env.${nodeEnv}`,
+          `.env.${nodeEnv}.local`,
+        ];
+      }
+
+      if (!this.#cached) {
```

Comment:
We should check cache using webpack API, there is a simple example usage cache API https://github.com/webpack-contrib/json-minimizer-webpack-plugin/blob/master/src/index.js#L122

Reply:
i.e. `await cache.getPromise("dotenv-webpack-plugin-data", null)` should return value, so if values is defined - we have cache and we don't need `this.#cached` at all

Reply:
somehow if i am adding await before the first execution and going in an asynchronous manner in the first run, i am facing some error. Like the code isnt executing correctly in the test created, that is why the first execution is kept synchronous.
can you just guide on this once